File: windows/TrikiMusicController.Windows/Core/TrikiButtonInterpreter.cs
using TrikiMusicController_Windows.Models;

namespace TrikiMusicController_Windows.Core;

public sealed class TrikiButtonInterpreter
{
    private const int MinimumProtocolObservations = 12;
    private const int MinimumRepeatedStatusRun = 4;
    private const long DebounceNanos = 18_000_000;
    private const long MinimumClickPressNanos = 25_000_000;
    private const long MaximumClickPressNanos = 2_000_000_000;
    private const long MultiClickTimeoutNanos = 450_000_000;
    private const long MaximumStreamGapNanos = 300_000_000;
    private long? _lastTimestampNanos;
    private int? _observedStatus;
    private int _observedRunLength;
    private int _longestObservedRun;
    private int _observationCount;
    private bool _stablePressed;
    private bool _candidatePressed;
    private long _candidateSinceNanos;
    private long? _pressedAtNanos;
    private int _pendingClickCount;
    private long? _clickDeadlineNanos;
    private bool _currentHoldConsumed;

    public TrikiButtonProtocolMode ProtocolMode { get; private set; } = TrikiButtonProtocolMode.Unknown;
    public bool IsPressed => ProtocolMode == TrikiButtonProtocolMode.ButtonFlag && _stablePressed;
    public bool ShouldSuppressMotionControl => ProtocolMode == TrikiButtonProtocolMode.ButtonFlag &&
        (_stablePressed || _candidatePressed || _pendingClickCount > 0);

    public bool ConsumeCurrentHold()
    {
        if (!IsPressed) return false;
        _currentHoldConsumed = true;
        _pendingClickCount = 0;
        _clickDeadlineNanos = null;
        return true;
    }

    public bool CheckAndConsumeHoldDuration(long now, long requiredDurationNanos)
    {
        if (!IsPressed || _currentHoldConsumed || _pressedAtNanos is not long pressedAt)
            return false;

        if (now - pressedAt >= requiredDurationNanos)
        {
            ConsumeCurrentHold();
            return true;
        }
        return false;
    }

    public void Reset()
    {
        ProtocolMode = TrikiButtonProtocolMode.Unknown;
        _lastTimestampNanos = null;
        _observedStatus = null;
        _observedRunLength = 0;
        _longestObservedRun = 0;
        _observationCount = 0;
        ClearInteraction();
    }

    public ButtonClickEvent? Process(TrikiSensorData sample)
    {
        var now = sample.TimestampNanos;
        if (_lastTimestampNanos is long previous && (now <= previous || now - previous > MaximumStreamGapNanos)) Reset();
        _lastTimestampNanos = now;

        if (ProtocolMode == TrikiButtonProtocolMode.Unknown)
        {
            ObserveProtocol(sample.Status, now);
            return null;
        }
        if (ProtocolMode == TrikiButtonProtocolMode.SequenceCounter) return null;
        if (sample.Status is < 0 or > 1)
        {
            ProtocolMode = TrikiButtonProtocolMode.SequenceCounter;
            ClearInteraction();
            return null;
        }
        return ProcessButtonState(sample.Status == 1, now);
    }

    private void ObserveProtocol(int status, long now)
    {
        if (status is < 0 or > 1)
        {
            ProtocolMode = TrikiButtonProtocolMode.SequenceCounter;
            ClearInteraction();
            return;
        }

        _observationCount++;
        if (_observedStatus == status) _observedRunLength++;
        else
        {
            _observedStatus = status;
            _observedRunLength = 1;
        }
        _longestObservedRun = Math.Max(_longestObservedRun, _observedRunLength);
        if (_observationCount < MinimumProtocolObservations || _longestObservedRun < MinimumRepeatedStatusRun) return;

        ProtocolMode = TrikiButtonProtocolMode.ButtonFlag;
        _stablePressed = status == 1;
        _candidatePressed = _stablePressed;
        _candidateSinceNanos = now;
        _pressedAtNanos = _stablePressed ? now : null;
        _pendingClickCount = 0;
        _clickDeadlineNanos = null;
        _currentHoldConsumed = false;
    }

    private ButtonClickEvent? ProcessButtonState(bool rawPressed, long now)
    {
        var completed = FinalizeExpiredSequence(rawPressed, now);
        if (rawPressed != _candidatePressed)
        {
            _candidatePressed = rawPressed;
            _candidateSinceNanos = now;
        }
        if (_candidatePressed == _stablePressed || now - _candidateSinceNanos < DebounceNanos) return completed;

        _stablePressed = _candidatePressed;
        if (_stablePressed)
        {
            _pressedAtNanos = now;
            _currentHoldConsumed = false;
        }
        else
        {
            completed ??= RegisterRelease(now);
        }
        return completed;
    }

    private ButtonClickEvent? FinalizeExpiredSequence(bool rawPressed, long now) =>
        _clickDeadlineNanos is long deadline && now >= deadline && !_stablePressed && !rawPressed
            ? CompletePendingSequence(now)
            : null;

    private ButtonClickEvent? RegisterRelease(long now)
    {
        var pressedAt = _pressedAtNanos;
        _pressedAtNanos = null;
        if (pressedAt is null) return null;
        if (_currentHoldConsumed)
        {
            _currentHoldConsumed = false;
            _pendingClickCount = 0;
            _clickDeadlineNanos = null;
            return null;
        }

        var duration = now - pressedAt.Value;
        if (duration is < MinimumClickPressNanos or > MaximumClickPressNanos)
        {
            _pendingClickCount = 0;
            _clickDeadlineNanos = null;
            return null;
        }
        _pendingClickCount++;
        if (_pendingClickCount >= 3) return CompletePendingSequence(now);
        _clickDeadlineNanos = now + MultiClickTimeoutNanos;
        return null;
    }

    private ButtonClickEvent? CompletePendingSequence(long now)
    {
        var type = _pendingClickCount switch
        {
            1 => ButtonClickType.Single,
            2 => ButtonClickType.Double,
            3 => ButtonClickType.Triple,
            _ => (ButtonClickType?)null,
        };
        _pendingClickCount = 0;
        _clickDeadlineNanos = null;
        _currentHoldConsumed = false;
        return type is ButtonClickType value ? new ButtonClickEvent(value, now) : null;
    }

    private void ClearInteraction()
    {
        _stablePressed = false;
        _candidatePressed = false;
        _candidateSinceNanos = 0;
        _pressedAtNanos = null;
        _pendingClickCount = 0;
        _clickDeadlineNanos = null;
        _currentHoldConsumed = false;
    }
}

File: app/src/main/java/pl/trikimusic/controller/core/bluetooth/TrikiButtonInterpreter.kt
package pl.trikimusic.controller.core.bluetooth

import pl.trikimusic.controller.domain.model.ButtonClickEvent
import pl.trikimusic.controller.domain.model.ButtonClickType
import pl.trikimusic.controller.domain.model.TrikiSensorData

/**
 * Interprets byte 1 only after identifying the connected firmware variant.
 * Some controllers expose a 0/1 button flag, while others put a rolling 0..15
 * packet id in the same byte. Waiting for a repeated 0/1 run prevents packet ids
 * from becoming phantom media clicks.
 */
class TrikiButtonInterpreter {
    var protocolMode: TrikiButtonProtocolMode = TrikiButtonProtocolMode.UNKNOWN
        private set

    val shouldSuppressMotionControl: Boolean
        get() = protocolMode == TrikiButtonProtocolMode.BUTTON_FLAG &&
            (stablePressed || candidatePressed || pendingClickCount > 0)

    val isPressed: Boolean
        get() = protocolMode == TrikiButtonProtocolMode.BUTTON_FLAG && stablePressed

    private var lastTimestampNanos: Long? = null
    private var observedStatus: Int? = null
    private var observedRunLength = 0
    private var longestObservedRun = 0
    private var observationCount = 0

    private var stablePressed = false
    private var candidatePressed = false
    private var candidateSinceNanos = 0L
    private var pressedAtNanos: Long? = null
    private var pendingClickCount = 0
    private var clickDeadlineNanos: Long? = null
    private var currentHoldConsumed = false

    fun consumeCurrentHold(): Boolean {
        if (!isPressed) return false
        currentHoldConsumed = true
        pendingClickCount = 0
        clickDeadlineNanos = null
        return true
    }

    fun checkAndConsumeHoldDuration(now: Long, requiredDurationNanos: Long): Boolean {
        val pressedAt = pressedAtNanos
        if (!isPressed || currentHoldConsumed || pressedAt == null) {
            return false
        }
        if (now - pressedAt >= requiredDurationNanos) {
            consumeCurrentHold()
            return true
        }
        return false
    }

    fun reset() {
        protocolMode = TrikiButtonProtocolMode.UNKNOWN
        lastTimestampNanos = null
        observedStatus = null
        observedRunLength = 0
        longestObservedRun = 0
        observationCount = 0
        clearInteraction()
    }

    fun process(sample: TrikiSensorData): ButtonClickEvent? {
        val now = sample.timestampNanos
        val previousTimestamp = lastTimestampNanos
        if (
            previousTimestamp != null &&
            (now <= previousTimestamp || now - previousTimestamp > MAX_STREAM_GAP_NANOS)
        ) {
            reset()
        }
        lastTimestampNanos = now

        if (protocolMode == TrikiButtonProtocolMode.UNKNOWN) {
            observeProtocol(sample.status, now)
            return null
        }
        if (protocolMode == TrikiButtonProtocolMode.SEQUENCE_COUNTER) return null

        if (sample.status !in BUTTON_STATUS_RANGE) {
            protocolMode = TrikiButtonProtocolMode.SEQUENCE_COUNTER
            clearInteraction()
            return null
        }

        return processButtonState(sample.status == PRESSED_STATUS, now)
    }

    private fun observeProtocol(status: Int, now: Long) {
        if (status !in BUTTON_STATUS_RANGE) {
            protocolMode = TrikiButtonProtocolMode.SEQUENCE_COUNTER
            clearInteraction()
            return
        }

        observationCount++
        if (observedStatus == status) {
            observedRunLength++
        } else {
            observedStatus = status
            observedRunLength = 1
        }
        longestObservedRun = maxOf(longestObservedRun, observedRunLength)

        if (
            observationCount >= MIN_PROTOCOL_OBSERVATIONS &&
            longestObservedRun >= MIN_REPEATED_STATUS_RUN
        ) {
            protocolMode = TrikiButtonProtocolMode.BUTTON_FLAG
            stablePressed = status == PRESSED_STATUS
            candidatePressed = stablePressed
            candidateSinceNanos = now
            pressedAtNanos = now.takeIf { stablePressed }
            pendingClickCount = 0
            clickDeadlineNanos = null
            currentHoldConsumed = false
        }
    }

    private fun processButtonState(rawPressed: Boolean, now: Long): ButtonClickEvent? {
        var completed = finalizeExpiredSequence(rawPressed, now)

        if (rawPressed != candidatePressed) {
            candidatePressed = rawPressed
            candidateSinceNanos = now
        }

        if (
            candidatePressed != stablePressed &&
            now - candidateSinceNanos >= DEBOUNCE_NANOS
        ) {
            stablePressed = candidatePressed
            if (stablePressed) {
                pressedAtNanos = now
                currentHoldConsumed = false
            } else {
                val releasedEvent = registerRelease(now)
                if (completed == null) completed = releasedEvent
            }
        }

        return completed
    }

    private fun finalizeExpiredSequence(rawPressed: Boolean, now: Long): ButtonClickEvent? {
        val deadline = clickDeadlineNanos ?: return null
        if (now < deadline || stablePressed || rawPressed) return null
        return completePendingSequence(now)
    }

    private fun registerRelease(now: Long): ButtonClickEvent? {
        val pressedAt = pressedAtNanos
        pressedAtNanos = null
        if (pressedAt == null) return null

        if (currentHoldConsumed) {
            currentHoldConsumed = false
            pendingClickCount = 0
            clickDeadlineNanos = null
            return null
        }

        val duration = now - pressedAt
        if (duration !in MIN_CLICK_PRESS_NANOS..MAX_CLICK_PRESS_NANOS) {
            pendingClickCount = 0
            clickDeadlineNanos = null
            return null
        }

        pendingClickCount++
        if (pendingClickCount >= ButtonClickType.TRIPLE.clickCount) {
            return completePendingSequence(now)
        }
        clickDeadlineNanos = now + MULTI_CLICK_TIMEOUT_NANOS
        return null
    }

    private fun completePendingSequence(timestampNanos: Long): ButtonClickEvent? {
        val type = when (pendingClickCount) {
            ButtonClickType.SINGLE.clickCount -> ButtonClickType.SINGLE
            ButtonClickType.DOUBLE.clickCount -> ButtonClickType.DOUBLE
            ButtonClickType.TRIPLE.clickCount -> ButtonClickType.TRIPLE
            else -> null
        }
        pendingClickCount = 0
        clickDeadlineNanos = null
        currentHoldConsumed = false
        return type?.let { ButtonClickEvent(it, timestampNanos) }
    }

    private fun clearInteraction() {
        stablePressed = false
        candidatePressed = false
        candidateSinceNanos = 0L
        pressedAtNanos = null
        pendingClickCount = 0
        clickDeadlineNanos = null
        currentHoldConsumed = false
    }

    private companion object {
        val BUTTON_STATUS_RANGE = 0..1
        const val PRESSED_STATUS = 1
        const val MIN_PROTOCOL_OBSERVATIONS = 12
        const val MIN_REPEATED_STATUS_RUN = 4
        const val DEBOUNCE_NANOS = 18_000_000L
        const val MIN_CLICK_PRESS_NANOS = 25_000_000L
        const val MAX_CLICK_PRESS_NANOS = 2_000_000_000L
        const val MULTI_CLICK_TIMEOUT_NANOS = 450_000_000L
        const val MAX_STREAM_GAP_NANOS = 300_000_000L
    }
}

enum class TrikiButtonProtocolMode(val displayName: String) {
    UNKNOWN("Rozpoznawanie"),
    BUTTON_FLAG("Przycisk 0/1"),
    SEQUENCE_COUNTER("Licznik ramek"),
}
